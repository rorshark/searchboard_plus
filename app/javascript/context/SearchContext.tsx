import React from 'react'
import { useSearch, SearchAPI, SearchParams } from '../hooks/useSearch'

export const defaultSearchApi: SearchAPI = {
  error: false,
  loading: true,
  customers: [],
  search: '',
  setSearch: (_) => {},
  companies: [],
  selectedCompany: null,
  setSelectedCompany: (_) => {},
  fetchCustomers: async () => {}
}

export const SearchContext: React.Context<SearchAPI> = React.createContext(defaultSearchApi)

export const SearchProvider: React.FC<{
  initialSearchParams: SearchParams
}> = ({ initialSearchParams, children }) => {
  const search = useSearch(initialSearchParams)

  return (
    <SearchContext.Provider value={search}>
      {children}
    </SearchContext.Provider>
  )
}
