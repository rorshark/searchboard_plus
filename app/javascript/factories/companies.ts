import { Factory } from 'fishery'
import Faker from 'faker'
import { Company } from '../types/companies'

export const company = Factory.define<Company>(() => ({
  companyName: Faker.company.companyName()
}))
