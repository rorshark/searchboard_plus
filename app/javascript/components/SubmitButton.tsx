import React, { useContext } from 'react'
import { SearchContext } from '../context/SearchContext'
import { Button } from '@material-ui/core'

const SubmitButton: React.FC = () => {
  const { fetchCustomers } = useContext(SearchContext)

  return (
    <Button 
      type="submit" 
      color="primary"
      variant="contained"
      data-testid="submit-button" 
      onClick={() => fetchCustomers()}>
      Search
    </Button>
  )
}

export default SubmitButton
