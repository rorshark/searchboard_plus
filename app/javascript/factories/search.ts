import { Factory } from 'fishery'
import { SearchAPI, SearchParams } from '../hooks/useSearch'
import { defaultSearchApi } from '../context/SearchContext'

export const searchApi = Factory.define<SearchAPI>(() => defaultSearchApi)

export const searchParams = Factory.define<SearchParams>(() => ({
  search: '',
  filter_by_company_name: ''
}))
