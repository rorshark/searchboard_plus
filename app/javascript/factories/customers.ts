import { Factory } from 'fishery'
import Faker from 'faker'
import { Customer } from '../types/customers'
import { company } from './companies'

export const customer = Factory.define<Customer>(() => ({
  firstName: Faker.name.firstName(),
  lastName: Faker.name.lastName(),
  company: company.build()
}));
