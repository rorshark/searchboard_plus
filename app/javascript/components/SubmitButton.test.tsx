import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import { SearchContext } from '../context/SearchContext'
import { searchApi } from '../factories/search'
import SubmitButton from './SubmitButton'

test('it renders a submit button', () => {
  const context = searchApi.build()

  render(
    <SearchContext.Provider value={context}>
      <SubmitButton />
    </SearchContext.Provider>
  )

  const element = screen.getByTestId('submit-button')

  expect(element).toMatchSnapshot()
})

test('it fetches customers when clicked', () => {
  const fetchCustomers = jest.fn()

  const context = searchApi.build({
    fetchCustomers
  })

  render(
    <SearchContext.Provider value={context}>
      <SubmitButton />
    </SearchContext.Provider>
  )

  const element = screen.getByTestId('submit-button')

  fireEvent.click(element)

  expect(fetchCustomers).toHaveBeenCalled()
})
