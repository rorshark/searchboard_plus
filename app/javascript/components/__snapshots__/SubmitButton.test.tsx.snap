// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders a submit button 1`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
  data-testid="submit-button"
  tabindex="0"
  type="submit"
>
  <span
    class="MuiButton-label"
  >
    Search
  </span>
  <span
    class="MuiTouchRipple-root"
  />
</button>
`;
